File: agent-driven-architecture-main/examples/kotlin/block/triage/src/main/kotlin/adr/blocks/triage/Project.kt
// ── blocks/triage/project — TWO pure projections of the SAME slice ─────────
// slice → view      what a human reads    (6.9, the Presenter)
// slice → lines     what the reasoner reads (G15, the third projection)
//
// They live on one CONSTRUCTED type because they are the same kind of thing: total,
// pure functions of committed state, with no clock, no I/O and no accumulator. The
// context lines are BOUNDED by declaration, so the reasoner's input does not
// grow with session length.
//
// THE BOUND IS DECLARED IN THE SPINE AND PASSED IN (docs/DECISIONS.md:174). This file
// used to argue the opposite — that injecting it would make it "a per-app knob and
// delete the lesson" — and the argument was wrong on its own terms: the lesson is that
// the projection is bounded BY DECLARATION rather than by discipline, and a defaulted
// parameter naming `MAX_CONTEXT_LINES_PER_BLOCK` declares it exactly as loudly. What
// the constant additionally did was make the bound UNCHECKABLE: it stamped the
// committed digest and re-derived it, so moving it left both gates green. The default
// is now pinned to its literal by a test, and a timeline re-derived under a different
// window diverges — neither of which the welded constant could express.

package adr.blocks.triage

import adr.contract.TriageResult.Priority
import adr.spine.pure.MAX_CONTEXT_LINES_PER_BLOCK

data class TicketRow(
    val ticket: String,
    val body: String,
    /** Pre-computed here, not in the view. */
    val badge: String,
)

data class TriageView(val rows: List<TicketRow>)

internal class TriageProjection {

    fun view(slice: TriageSlice): TriageView = TriageView(
        rows = slice.tickets.values.map { ticket ->
            TicketRow(
                ticket = ticket.id.value,
                body = ticket.body,
                badge = (slice.priority[ticket.id] ?: Priority.Normal).name.uppercase(),
            )
        },
    )

    fun contextLines(slice: TriageSlice, maxLines: Int = MAX_CONTEXT_LINES_PER_BLOCK): List<String> =
        slice.tickets.values
            .take(maxLines)
            .map { "ticket ${it.id.value} [${(slice.priority[it.id] ?: Priority.Normal).name}] ${it.body}" }
}

File: agent-driven-architecture-main/examples/typescript/src/blocks/triage/project.ts
// ── blocks/triage/project — TWO pure projections of the same slice ─────────
// slice → view      the Presenter (6.9): every presentational flag pre-decided
// slice → context   the reasoner's input (6.11/G15): a BOUNDED digest
//
// They live in one file because they are the same kind of thing — pure
// projections of committed State — and they must be read side by side, so that
// "what the user sees" and "what the model sees" are never accidentally
// different derivations of the same fact.

import { bounded, MAX_CONTEXT_LINES_PER_BLOCK } from "@adr/spine/pure/context";
import type { Priority } from "./contract";
import type { TriageSlice } from "./slice";

export interface TriageRow {
  readonly ticket: string;
  readonly badge: string;
  readonly priority: Priority;
}

export interface TriageView {
  readonly rows: readonly TriageRow[];
}

export function triageView(slice: TriageSlice): TriageView {
  const rows = [...slice.tickets.values()].map((t) => {
    const priority = slice.priority.get(t.id) ?? "Normal";
    return { ticket: t.id, badge: priority.toUpperCase(), priority };
  });
  return { rows };
}

/** `max` is the ROOT'S window, defaulted to the spine's shipped one
 *  (docs/DECISIONS.md:174): the block still declares a bound, it just no longer
 *  decides the number. */
export function triageContextLines(
  slice: TriageSlice,
  max: number = MAX_CONTEXT_LINES_PER_BLOCK,
): readonly string[] {
  const lines = [...slice.tickets.values()].map(
    (t) => `ticket ${t.id} [${slice.priority.get(t.id) ?? "Normal"}]: ${t.body}`,
  );
  return bounded(lines, max);
}
